shared/src/types/token.rs

```diff
@@ -279,20 +296,58 @@ pub fn is_any_token_balance_key(key: &Key) -> Option<&Address> {
     }
 }
 
-/// Check if the given storage key is non-owner's balance key. If it is, returns
-/// the address.
-pub fn is_non_owner_balance_key(key: &Key) -> Option<&Address> {
-    match &key.segments[..] {
-        [
-            DbKeySeg::AddressSeg(_),
-            DbKeySeg::StringSeg(key),
-            DbKeySeg::AddressSeg(owner),
-        ] if key == BALANCE_STORAGE_KEY => match owner {
-            Address::Internal(InternalAddress::IbcEscrow(_))
-            | Address::Internal(InternalAddress::IbcBurn)
-            | Address::Internal(InternalAddress::IbcMint) => Some(owner),
-            _ => None,
-        },
+/// Check if the given storage key is multitoken balance key for the given
+/// token. If it is, returns the sub prefix and the owner.
+pub fn is_multitoken_balance_key<'a>(
+    token_addr: &Address,
+    key: &'a Key,
+) -> Option<(Key, &'a Address)> {
+    match key.segments.first() {
+        Some(DbKeySeg::AddressSeg(addr)) if addr == token_addr => {}
+        _ => return None,
+    }
+    let len = key.segments.len();
+    match key.get_at(len - 2) {
+        Some(DbKeySeg::StringSeg(balance))
+            if balance == BALANCE_STORAGE_KEY => {}
+        _ => return None,
+    }
+    match key.segments.last() {
+        Some(DbKeySeg::AddressSeg(owner)) => {
+            let sub_prefix = Key {
+                segments: key.segments[1..(len - 2)].to_vec(),
+            };
+            Some((sub_prefix, owner))
+        }
+        _ => None,
+    }
+}
+
+/// Check if the given storage key is multitoken balance key for unspecified
+/// token. If it is, returns the sub prefix and the owner.
+pub fn is_any_multitoken_balance_key(key: &Key) -> Option<(Key, &Address)> {
+    match key.segments.first() {
+        Some(DbKeySeg::AddressSeg(_)) => {}
+        _ => return None,
+    }
+    let len = key.segments.len();
+    if len <= 3 {
+        return None;
+    }
+    match key.get_at(len - 2) {
+        Some(DbKeySeg::StringSeg(balance))
+            if balance == BALANCE_STORAGE_KEY =>
+        {
+            match key.segments.last() {
+                Some(DbKeySeg::AddressSeg(owner)) => {
+                    let sub_prefix = Key {
+                        segments: key.segments[1..(len - 2)].to_vec(),
+                    };
+                    Some((sub_prefix, owner))
+                }
+                _ => None,
+            }
+        }
         _ => None,
     }
 }
```

Comment:
Can `is_multitoken_balance_key`/`is_any_multitoken_balance_key` share some code? These would be useful to have tests for as well (asserting what multitoken keys look like)